src/components/Ruler/index.tsx

```diff
@@ -0,0 +1,149 @@
+import * as React from 'react';
+import RulerTooltip from '../RulerTooltip';
+import { ItemId } from '../../external';
+
+export interface RulerConfig {
+  visible: boolean;
+  timeLabel: (data: RulerPoint) => string;
+  yLabel: (data: RulerPoint) => string;
+  getTimeLabelPosition?: (
+    defaultPosition: number,
+    measurements: {
+      height: number;
+      labelHeight: number;
+      timeLabelMargin: number;
+    }
+  ) => number;
+}
+
+export interface RulerPoint {
+  id?: ItemId;
+  name: string;
+  color: string;
+  timestamp: number;
+  value: number | string;
+  x: number;
+  y: number;
+}
+
+export interface Props {
+  chartWidth: number;
+  chartHeight: number;
+  points: RulerPoint[];
+  ruler: RulerConfig;
+}
+
+const labelHeight = 24;
+
+const calculateY = (points: RulerPoint[], yTooltipPosition: number) => {
+  const pointsObject: { [name: string]: number } = {};
+
+  const pointsSorted = [...points].sort((a, b) => a.y - b.y);
+
+  let prevPoint = undefined;
+  for (let i = 0; i < pointsSorted.length; i += 1) {
+    const realSpace = yTooltipPosition - pointsSorted[i].y;
+    const neededSpace = (pointsSorted.length - i) * labelHeight;
+
+    // move point upper if no space until the limit available
+    if (realSpace < neededSpace) {
+      const pointPos = yTooltipPosition - neededSpace;
+      // if no upper room available stick it to 0
+      prevPoint = pointPos >= 0 ? pointPos : 0;
+      // move point lower if it overlaps the previous one
+    } else if (
+      prevPoint !== undefined &&
+      prevPoint + labelHeight > pointsSorted[i].y
+    ) {
+      prevPoint += labelHeight;
+    } else {
+      prevPoint = pointsSorted[i].y;
+    }
+    pointsObject[pointsSorted[i].name] = prevPoint;
+  }
+
+  return pointsObject;
+};
+
+const Ruler: React.FunctionComponent<Props> = ({
+  ruler,
+  points,
+  chartHeight,
+  chartWidth,
+}) => {
+  const {
+    visible = true,
+    timeLabel = ({ timestamp }: RulerPoint) => String(new Date(timestamp)),
+    yLabel = ({ value }: RulerPoint) => String(value),
+    getTimeLabelPosition = (defaultTimeTooltipPosition: number) =>
+      defaultTimeTooltipPosition,
+  } = ruler;
+  if (!visible) {
+    return null;
+  }
+
+  const xScale = (x: number) => x;
+
+  const timeLabelMargin = 15;
```

Comment:
5 -> 15, just in case that was not intentional

Reply:
Good catch. I remember why I made this change -- we actually needed to double it to get the math right (otherwise it exceeds the boundary of the SVG)